File: javatypes.py
# -*- coding: utf-8 -*-
import re

_PRIMS = {
    "V": "void", "Z": "boolean", "B": "byte", "C": "char",
    "S": "short", "I": "int", "J": "long", "F": "float", "D": "double",
}


def dotted_from_internal(internal):
    """pkg/Outer$Inner -> pkg.Outer.Inner (валидная Java-ссылка на класс).
    Числовые сегменты (анонимные классы, Outer$1) оставляем приклеенными
    через $ - в исходном Java-коде сослаться на анонимный класс по имени
    синтаксически невозможно, точечная форма тут была бы ещё более неверной."""
    pkg, _, simple = internal.rpartition("/")
    segs = simple.split("$")
    pieces = [segs[0]]
    for s in segs[1:]:
        if s.isdigit():
            pieces[-1] = pieces[-1] + "$" + s
        else:
            pieces.append(s)
    result = ".".join(pieces)
    dotted_pkg = pkg.replace("/", ".")
    return f"{dotted_pkg}.{result}" if dotted_pkg else result


def _parse_one_type(desc, i):
    c = desc[i]
    if c in _PRIMS:
        return _PRIMS[c], i + 1
    if c == "L":
        j = desc.index(";", i)
        internal = desc[i + 1:j]
        return dotted_from_internal(internal), j + 1
    if c == "[":
        inner, ni = _parse_one_type(desc, i + 1)
        return inner + "[]", ni
    raise ValueError(f"Плохой дескриптор типа: {desc!r} at {i}")


TYPE_MARK_OPEN = "\x01"
TYPE_MARK_CLOSE = "\x02"


def mark_type(dotted):
    """Оборачивает dotted-имя типа в маркер-плейсхолдер вместо немедленного
    сведения к simple-имени. Нужно для отложенного (второй проход) решения
    об imports/FQN - см. resolve_type_markers() ниже и HANDOFF_STATUS_RU.md,
    пункт B.1 (коллизии simple-имён классов из разных пакетов, напр.
    org.bukkit.ChatColor vs net.md_5.bungee.api.ChatColor).
    Возвращает текст с маркером вокруг БАЗОВОГО типа (без "[]" - суффикс
    массива дописывается снаружи маркера как есть)."""
    if dotted is None or dotted == "":
        return "Object" if dotted is None else dotted
    arr = ""
    base = dotted
    while base.endswith("[]"):
        arr += "[]"
        base = base[:-2]
    return f"{TYPE_MARK_OPEN}{base}{TYPE_MARK_CLOSE}{arr}"


def resolve_type_markers(text, losers):
    """Финальный проход по уже собранному тексту класса: заменяет каждый
    mark_type()-маркер на:
      - simple-имя (обычный случай, тип импортирован под коротким именем)
      - полное dotted-имя (FQN), если тип входит в `losers` - т.е. проиграл
        коллизию simple-имён другому типу, который встретился в файле раньше
        (первый встреченный тип коллизии сохраняет `import X;` + short name,
        остальные печатаются как FQN и НЕ импортируются - см. render_class)."""
    out = []
    i = 0
    n = len(text)
    while i < n:
        ch = text[i]
        if ch == TYPE_MARK_OPEN:
            j = text.index(TYPE_MARK_CLOSE, i + 1)
            dotted = text[i + 1:j]
            if dotted in losers:
                out.append(dotted)
            else:
                out.append(dotted.rsplit(".", 1)[-1] if "." in dotted else dotted)
            i = j + 1
        else:
            out.append(ch)
            i += 1
    return "".join(out)


def field_descriptor_to_java(desc):
    t, _ = _parse_one_type(desc, 0)
    return t


def method_descriptor_to_java(desc):
    """returns (return_type_str, [param_type_str, ...])"""
    assert desc.startswith("(")
    i = 1
    params = []
    while desc[i] != ")":
        t, i = _parse_one_type(desc, i)
        params.append(t)
    ret, _ = _parse_one_type(desc, i + 1)
    return ret, params


# ---------------- obfuscation heuristics ----------------

VOWELS = set("aeiou")

JAVA_KEYWORDS = {
    "abstract", "assert", "boolean", "break", "byte", "case", "catch", "char",
    "class", "const", "continue", "default", "do", "double", "else", "enum",
    "extends", "final", "finally", "float", "for", "goto", "if", "implements",
    "import", "instanceof", "int", "interface", "long", "native", "new",
    "package", "private", "protected", "public", "return", "short", "static",
    "strictfp", "super", "switch", "synchronized", "this", "throw", "throws",
    "transient", "try", "void", "volatile", "while", "true", "false", "null",
    "var", "record", "yield", "sealed", "permits",
}

_IDENT_RE = re.compile(r"^[A-Za-z_$][A-Za-z0-9_$]*$")


def is_safe_local_name(name):
    """Годится ли имя локальной переменной/параметра из LocalVariableTable для
    прямой подстановки в Java-текст: валидный идентификатор и не
    зарезервированное слово (обфускатор иногда генерирует отладочную таблицу
    с мусорными/невалидными именами - подстраховываемся)."""
    return bool(name) and bool(_IDENT_RE.match(name)) and name not in JAVA_KEYWORDS


def _consonant_run(name_lower):
    run = 0
    best = 0
    for ch in name_lower:
        if ch.isalpha() and ch not in VOWELS:
            run += 1
            best = max(best, run)
        else:
            run = 0
    return best


def looks_obfuscated(name, kind="class"):
    """
    Эвристика: похоже ли имя на сгенерированное обфускатором (случайный набор букв),
    а не на осмысленный человеческий идентификатор.
    kind: 'class' | 'method' | 'field' | 'package'

    ВАЖНО: намеренно консервативная эвристика (см. фидбек пользователя -
    реальные разработчики часто используют короткие/"дурацкие"/сокращённые
    имена, и старые пороги слишком часто путали их с обфускацией, портя
    вполне читаемый код). По умолчанию переименовываем только то, что
    выглядит как почти гарантированная обфускация (одна буква, либо длинный
    плоский нижнерегистровый набор букв без единой гласной закономерности) -
    остальное лучше оставить как есть, даже если оно кажется "странным".
    """
    if name is None:
        return False
    if name in ("<init>", "<clinit>"):
        return False

    base = name.rsplit("$", 1)[-1]  # для внутренних классов Foo$1 смотрим на последнюю часть

    # Однобуквенные top-level имена - классический сигнатурный признак
    # обфускатора (a.class, b()); всё остальное (в т.ч. 2-буквенные - могут
    # быть осмысленной аббревиатурой вроде "Io"/"Db") не флагаем по длине.
    if kind in ("class", "method", "field") and len(base) == 1:
        return True

    lower = base.lower()
    letters_only = re.sub(r"[^a-z]", "", lower)
    if len(letters_only) < 12:
        return False  # короткое/среднее имя - не флагаем, слишком велик риск ложного срабатывания

    has_upper = any(c.isupper() for c in base)
    has_digit = any(c.isdigit() for c in base)
    has_underscore = "_" in base

    vowel_ratio = sum(1 for c in letters_only if c in VOWELS) / len(letters_only)
    max_run = _consonant_run(letters_only)

    # Сигналы "случайности" - оставлены только КРАЙНИЕ случаи:
    flat_lower_long = (not has_upper) and (not has_digit) and (not has_underscore) and len(base) >= 12
    very_low_vowels = vowel_ratio < 0.20
    very_long_consonant_run = max_run >= 6

    if kind in ("class", "method", "field"):
        if flat_lower_long and (very_low_vowels or very_long_consonant_run):
            return True
        if flat_lower_long and len(base) >= 18:
            # очень длинные плоские нижнерегистровые идентификаторы без единой
            # заглавной буквы/цифры - почти наверняка обфускация, а не реальное
            # (пусть и длинное) человеческое имя
            return True

    if kind == "package":
        if flat_lower_long and len(base) >= 16:
            return True

    return False

File: resources/engine_cpp/src/javatypes.cpp
// javatypes.cpp - см. javatypes.hpp. 1:1 порт javatypes.py.
#include "javatypes.hpp"

#include <algorithm>
#include <cctype>
#include <map>
#include <stdexcept>

namespace nd {

namespace {

const std::map<char, std::string> PRIMS = {
    {'V', "void"}, {'Z', "boolean"}, {'B', "byte"}, {'C', "char"},
    {'S', "short"}, {'I', "int"}, {'J', "long"}, {'F', "float"}, {'D', "double"},
};

[[maybe_unused]] bool is_prim(char c) { return PRIMS.count(c) != 0; }

// Аналог str.split(sep) в Python (без maxsplit) - разбивает по КАЖДОМУ
// вхождению разделителя, сохраняя пустые сегменты.
std::vector<std::string> split_all(const std::string& s, char sep) {
    std::vector<std::string> out;
    size_t start = 0;
    for (size_t i = 0; i < s.size(); ++i) {
        if (s[i] == sep) {
            out.push_back(s.substr(start, i - start));
            start = i + 1;
        }
    }
    out.push_back(s.substr(start));
    return out;
}

bool is_ascii_digit_str(const std::string& s) {
    if (s.empty()) return false;
    for (char c : s) {
        if (!std::isdigit(static_cast<unsigned char>(c))) return false;
    }
    return true;
}

// rsplit(sep, 1)[-1] - последний сегмент после ПОСЛЕДНЕГО вхождения sep,
// или вся строка, если sep не встречается.
std::string last_segment_after(const std::string& s, char sep) {
    auto pos = s.rfind(sep);
    if (pos == std::string::npos) return s;
    return s.substr(pos + 1);
}

std::string join(const std::vector<std::string>& parts, const std::string& sep) {
    std::string out;
    for (size_t i = 0; i < parts.size(); ++i) {
        if (i) out += sep;
        out += parts[i];
    }
    return out;
}

}  // namespace

// ---------------- dotted_from_internal ----------------

std::string dotted_from_internal(const std::string& internal) {
    std::string pkg, simple;
    auto slash = internal.rfind('/');
    if (slash == std::string::npos) {
        pkg = "";
        simple = internal;
    } else {
        pkg = internal.substr(0, slash);
        simple = internal.substr(slash + 1);
    }
    auto segs = split_all(simple, '$');
    std::vector<std::string> pieces;
    pieces.push_back(segs[0]);
    for (size_t k = 1; k < segs.size(); ++k) {
        const std::string& s = segs[k];
        if (is_ascii_digit_str(s)) {
            pieces.back() += "$" + s;
        } else {
            pieces.push_back(s);
        }
    }
    std::string result = join(pieces, ".");
    std::string dotted_pkg = pkg;
    std::replace(dotted_pkg.begin(), dotted_pkg.end(), '/', '.');
    if (!dotted_pkg.empty()) return dotted_pkg + "." + result;
    return result;
}

// ---------------- descriptor parsing ----------------

std::pair<std::string, size_t> parse_one_type(const std::string& desc, size_t i) {
    if (i >= desc.size()) throw std::invalid_argument("Плохой дескриптор типа: '" + desc + "' at " + std::to_string(i));
    char c = desc[i];
    auto it = PRIMS.find(c);
    if (it != PRIMS.end()) return {it->second, i + 1};
    if (c == 'L') {
        auto j = desc.find(';', i);
        if (j == std::string::npos) throw std::invalid_argument("Плохой дескриптор типа: '" + desc + "' at " + std::to_string(i));
        std::string internal = desc.substr(i + 1, j - (i + 1));
        return {dotted_from_internal(internal), j + 1};
    }
    if (c == '[') {
        auto [inner, ni] = parse_one_type(desc, i + 1);
        return {inner + "[]", ni};
    }
    throw std::invalid_argument("Плохой дескриптор типа: '" + desc + "' at " + std::to_string(i));
}

std::string field_descriptor_to_java(const std::string& desc) {
    return parse_one_type(desc, 0).first;
}

std::pair<std::string, std::vector<std::string>> method_descriptor_to_java(const std::string& desc) {
    if (desc.empty() || desc[0] != '(') throw std::invalid_argument("method_descriptor_to_java: expected '(' at start");
    size_t i = 1;
    std::vector<std::string> params;
    while (i < desc.size() && desc[i] != ')') {
        auto [t, ni] = parse_one_type(desc, i);
        params.push_back(t);
        i = ni;
    }
    auto [ret, _] = parse_one_type(desc, i + 1);
    return {ret, params};
}

// ---------------- mark_type / resolve_type_markers ----------------

std::string mark_type(const std::optional<std::string>& dotted) {
    if (!dotted.has_value()) return "Object";
    if (dotted->empty()) return "";
    std::string arr;
    std::string base = *dotted;
    while (base.size() >= 2 && base.substr(base.size() - 2) == "[]") {
        arr += "[]";
        base = base.substr(0, base.size() - 2);
    }
    std::string out;
    out += TYPE_MARK_OPEN;
    out += base;
    out += TYPE_MARK_CLOSE;
    out += arr;
    return out;
}

std::string resolve_type_markers(const std::string& text, const std::unordered_set<std::string>& losers) {
    std::string out;
    size_t i = 0, n = text.size();
    while (i < n) {
        char ch = text[i];
        if (ch == TYPE_MARK_OPEN) {
            auto j = text.find(TYPE_MARK_CLOSE, i + 1);
            // Не должно случиться на корректно сформированном тексте
            // (каждый mark_type() всегда парный) - но на всякий случай, как
            // и оригинал (который тоже упал бы с ValueError на text.index()
            // при отсутствии закрывающего маркера), не глотаем ошибку молча.
            if (j == std::string::npos) throw std::invalid_argument("resolve_type_markers: unmatched TYPE_MARK_OPEN");
            std::string dotted = text.substr(i + 1, j - (i + 1));
            if (losers.count(dotted)) {
                out += dotted;
            } else {
                out += last_segment_after(dotted, '.');
            }
            i = j + 1;
        } else {
            out += ch;
            i += 1;
        }
    }
    return out;
}

// ---------------- Signature parsing ----------------

namespace {

struct SignatureParseError : std::runtime_error {
    explicit SignatureParseError(const std::string& m) : std::runtime_error(m) {}
};

class FieldSigParser {
public:
    explicit FieldSigParser(const std::string& s) : s_(s), i_(0), n_(s.size()) {}

    char peek() const {
        if (i_ >= n_) throw SignatureParseError("неожиданный конец сигнатуры");
        return s_[i_];
    }

    std::string parse_type() {
        char c = peek();
        if (c == 'L') return parse_class_type();
        if (c == 'T') return parse_type_var();
        if (c == '[') {
            i_ += 1;
            return parse_type() + "[]";
        }
        auto it = PRIMS.find(c);
        if (it != PRIMS.end()) {
            i_ += 1;
            return it->second;
        }
        if (c == '*') {
            i_ += 1;
            return "?";
        }
        if (c == '+' || c == '-') {
            char wc = c;
            i_ += 1;
            std::string inner = parse_type();
            return (wc == '+' ? std::string("? extends ") : std::string("? super ")) + inner;
        }
        throw SignatureParseError(std::string("неожиданный символ '") + c + "' в позиции " + std::to_string(i_));
    }

    std::string s_;
    size_t i_;
    size_t n_;

private:
    std::string parse_type_var() {
        i_ += 1;  // 'T'
        size_t start = i_;
        while (peek() != ';') i_ += 1;
        std::string name = s_.substr(start, i_ - start);
        i_ += 1;  // ';'
        return name;
    }

    std::string parse_class_type() {
        i_ += 1;  // 'L'
        size_t start = i_;
        while (peek() != '<' && peek() != ';' && peek() != '.') i_ += 1;
        std::string internal = s_.substr(start, i_ - start);
        std::string dotted = dotted_from_internal(internal);
        std::string result = last_segment_after(dotted, '.');
        if (peek() == '<') result += parse_type_args();
        while (i_ < n_ && peek() == '.') {
            i_ += 1;
            size_t istart = i_;
            while (peek() != '<' && peek() != ';' && peek() != '.') i_ += 1;
            std::string iname = s_.substr(istart, i_ - istart);
            result += "." + iname;
            if (peek() == '<') result += parse_type_args();
        }
        if (peek() != ';') throw SignatureParseError("ожидался ';' в конце ClassTypeSignature");
        i_ += 1;
        return result;
    }

    std::string parse_type_args() {
        i_ += 1;  // '<'
        std::vector<std::string> args;
        while (peek() != '>') args.push_back(parse_type());
        i_ += 1;  // '>'
        return "<" + join(args, ", ") + ">";
    }
};

class FullSigParser : public FieldSigParser {
public:
    explicit FullSigParser(const std::string& s) : FieldSigParser(s) {}

    std::string read_ident() {
        size_t start = i_;
        while (i_ < n_ && s_[i_] != ':' && s_[i_] != '<' && s_[i_] != ';' && s_[i_] != '.' &&
               s_[i_] != '[' && s_[i_] != '/' && s_[i_] != '(' && s_[i_] != ')' && s_[i_] != '^') {
            i_ += 1;
        }
        return s_.substr(start, i_ - start);
    }

    std::vector<std::pair<std::string, std::vector<std::string>>> parse_type_params() {
        std::vector<std::pair<std::string, std::vector<std::string>>> params;
        if (i_ >= n_ || peek() != '<') return params;
        i_ += 1;
        while (peek() != '>') {
            std::string name = read_ident();
            std::vector<std::string> bounds;
            if (peek() != ':') throw SignatureParseError("ожидался ':' после имени параметра типа");
            i_ += 1;  // первое ':' (ClassBound, может быть пустым)
            if (peek() != ':' && peek() != '>') bounds.push_back(parse_type());
            while (peek() == ':') {  // InterfaceBound*
                i_ += 1;
                bounds.push_back(parse_type());
            }
            if (bounds.size() == 1 && bounds[0] == "Object") bounds.clear();
            params.emplace_back(name, std::move(bounds));
        }
        i_ += 1;  // '>'
        return params;
    }
};

std::string format_type_params(const std::vector<std::pair<std::string, std::vector<std::string>>>& type_params) {
    if (type_params.empty()) return "";
    std::vector<std::string> parts;
    for (auto& [name, bounds] : type_params) {
        if (!bounds.empty()) {
            parts.push_back(name + " extends " + join(bounds, " & "));
        } else {
            parts.push_back(name);
        }
    }
    return "<" + join(parts, ", ") + ">";
}

}  // namespace

std::optional<std::string> parse_field_signature(const std::string& sig) {
    if (sig.empty()) return std::nullopt;
    try {
        FieldSigParser p(sig);
        std::string result = p.parse_type();
        if (p.i_ != p.n_) return std::nullopt;  // остался "хвост" - не доверяем
        return result;
    } catch (...) {
        return std::nullopt;  // SignatureParseError или любая другая ошибка разбора
    }
}

std::optional<MethodSignature> parse_method_signature(const std::string& sig) {
    if (sig.empty()) return std::nullopt;
    try {
        FullSigParser p(sig);
        auto type_params = p.parse_type_params();
        if (p.peek() != '(') return std::nullopt;
        p.i_ += 1;
        std::vector<std::string> param_types;
        while (p.peek() != ')') param_types.push_back(p.parse_type());
        p.i_ += 1;  // ')'
        std::string ret;
        if (p.peek() == 'V') {
            p.i_ += 1;
            ret = "void";
        } else {
            ret = p.parse_type();
        }
        while (p.i_ < p.n_ && p.peek() == '^') {
            p.i_ += 1;
            p.parse_type();  // typed throws - разбирается, но не используется
        }
        if (p.i_ != p.n_) return std::nullopt;
        MethodSignature ms;
        ms.type_params = format_type_params(type_params);
        ms.param_types = std::move(param_types);
        ms.return_type = ret;
        return ms;
    } catch (...) {
        return std::nullopt;
    }
}

std::optional<ClassSignature> parse_class_signature(const std::string& sig) {
    if (sig.empty()) return std::nullopt;
    try {
        FullSigParser p(sig);
        auto type_params = p.parse_type_params();
        std::string superclass = p.parse_type();
        std::vector<std::string> interfaces;
        while (p.i_ < p.n_) interfaces.push_back(p.parse_type());
        ClassSignature cs;
        cs.type_params = format_type_params(type_params);
        cs.superclass = superclass;
        cs.interfaces = std::move(interfaces);
        return cs;
    } catch (...) {
        return std::nullopt;
    }
}

// ---------------- obfuscation heuristics ----------------

namespace {
const std::unordered_set<std::string> JAVA_KEYWORDS = {
    "abstract", "assert", "boolean", "break", "byte", "case", "catch", "char",
    "class", "const", "continue", "default", "do", "double", "else", "enum",
    "extends", "final", "finally", "float", "for", "goto", "if", "implements",
    "import", "instanceof", "int", "interface", "long", "native", "new",
    "package", "private", "protected", "public", "return", "short", "static",
    "strictfp", "super", "switch", "synchronized", "this", "throw", "throws",
    "transient", "try", "void", "volatile", "while", "true", "false", "null",
    "var", "record", "yield", "sealed", "permits",
};

const std::unordered_set<char> VOWELS = {'a', 'e', 'i', 'o', 'u'};

bool is_ident_char_first(char c) {
    return std::isalpha(static_cast<unsigned char>(c)) || c == '_' || c == '$';
}
bool is_ident_char_rest(char c) {
    return std::isalnum(static_cast<unsigned char>(c)) || c == '_' || c == '$';
}

int consonant_run(const std::string& name_lower) {
    int run = 0, best = 0;
    for (char ch : name_lower) {
        if (std::isalpha(static_cast<unsigned char>(ch)) && !VOWELS.count(ch)) {
            run += 1;
            best = std::max(best, run);
        } else {
            run = 0;
        }
    }
    return best;
}

}  // namespace

bool is_safe_local_name(const std::string& name) {
    if (name.empty()) return false;
    if (!is_ident_char_first(name[0])) return false;
    for (size_t i = 1; i < name.size(); ++i) {
        if (!is_ident_char_rest(name[i])) return false;
    }
    return JAVA_KEYWORDS.count(name) == 0;
}

bool looks_obfuscated(const std::optional<std::string>& name_opt, const std::string& kind) {
    if (!name_opt.has_value()) return false;
    const std::string& name = *name_opt;
    if (name == "<init>" || name == "<clinit>") return false;

    std::string base;
    if (kind == "class") {
        base = last_segment_after(name, '$');
    } else {
        if (name.find('$') != std::string::npos) return false;
        base = name;
    }

    if ((kind == "class" || kind == "method" || kind == "field") && base.size() == 1) return true;

    std::string lower = base;
    std::transform(lower.begin(), lower.end(), lower.begin(),
                    [](unsigned char c) { return std::tolower(c); });
    std::string letters_only;
    for (char c : lower) {
        if (c >= 'a' && c <= 'z') letters_only += c;
    }
    if (letters_only.size() < 12) return false;

    bool has_upper = std::any_of(base.begin(), base.end(), [](unsigned char c) { return std::isupper(c); });
    bool has_digit = std::any_of(base.begin(), base.end(), [](unsigned char c) { return std::isdigit(c); });
    bool has_underscore = base.find('_') != std::string::npos;

    double vowel_ratio = 0.0;
    {
        size_t vc = 0;
        for (char c : letters_only) if (VOWELS.count(c)) vc += 1;
        vowel_ratio = static_cast<double>(vc) / static_cast<double>(letters_only.size());
    }
    int max_run = consonant_run(letters_only);

    bool flat_lower_long = (!has_upper) && (!has_digit) && (!has_underscore) && base.size() >= 12;
    bool very_low_vowels = vowel_ratio < 0.20;
    bool very_long_consonant_run = max_run >= 6;

    if (kind == "class" || kind == "method" || kind == "field") {
        if (flat_lower_long && (very_low_vowels || very_long_consonant_run)) return true;
        if (flat_lower_long && base.size() >= 18) return true;
    }
    if (kind == "package") {
        if (flat_lower_long && base.size() >= 16) return true;
    }
    return false;
}

}  // namespace nd
